optimum/intel/neural_compressor/modeling_decoder.py

```diff
@@ -0,0 +1,168 @@
+    embeddings).
+    """,
+    MODEL_START_DOCSTRING,
+)
+class INCModelForCausalLM(INCBaseModel, BaseModelForCausalLM):
+    export_feature = "text-generation"
+    auto_model_class = AutoModelForCausalLM
+
+    def __init__(
+        self,
+        model,
+        config: PretrainedConfig = None,
+        model_save_dir: Optional[Union[str, Path, TemporaryDirectory]] = None,
+        use_cache: bool = True,
+        **kwargs,
+    ):
+        super(INCBaseModel, self).__init__(
+            model=model, config=config, model_save_dir=model_save_dir, use_cache=use_cache, **kwargs
+        )
+        self.backend = getattr(config, "backend", None)
+
+    @add_start_docstrings_to_model_forward(
+        INPUTS_DOCSTRING.format("batch_size, sequence_length")
+        + TEXT_GENERATION_EXAMPLE.format(
+            processor_class=_TOKENIZER_FOR_DOC,
+            model_class="INCModelForCausalLM",
+            checkpoint="EleutherAI/gpt-j-6b",
+        )
+    )
+    def forward(
+        self,
+        input_ids: torch.LongTensor = None,
+        attention_mask: Optional[torch.FloatTensor] = None,
+        past_key_values: Optional[Tuple[Tuple[torch.Tensor]]] = None,
+        **kwargs,
+    ) -> CausalLMOutputWithPast:
+        inputs = {
+            "input_ids": input_ids,
+            "attention_mask": attention_mask,
+        }
+
+        nb_pkv = 2
+        num_layers = self.normalized_config.num_layers
+        first_token = False
+        if self.use_cache:
+            if past_key_values is None:
+                first_token = True
+                num_attention_heads = self.normalized_config.num_attention_heads
+                hidden_size = self.normalized_config.hidden_size
+                d_k = hidden_size // num_attention_heads
+
+                if self.config.model_type != "bloom":
+                    if self.backend == "neural_engine":
+                        new_shape = [input_ids.shape[0], 0, num_attention_heads, d_k]
+                    else:
+                        new_shape = [input_ids.shape[0], num_attention_heads, 0, d_k]
+                    empty_tensor = torch.empty(size=new_shape)
+                    past_key_values = tuple(tuple(empty_tensor for _ in range(nb_pkv)) for _ in range(num_layers))
+                    pkv = tuple(empty_tensor for _ in range(nb_pkv))
+                else:
+                    pkv = ()
+                    for nb_pkv in range(nb_pkv):
+                        if nb_pkv % 2 == 0:
+                            new_shape = [input_ids.shape[0] * num_attention_heads, d_k, 0]
+                        else:
+                            new_shape = [input_ids.shape[0] * num_attention_heads, 0, d_k]
+                        pkv = pkv + (torch.empty(size=new_shape),)
+                past_key_values = tuple(tuple(pkv) for _ in range(num_layers))
+
+            inputs["past_key_values"] = past_key_values
+        if self.backend == "neural_engine":
```

Comment:
why is it needed ?

Reply:
Neural engine need the inputs should be a tuple not a dict or tuple of tuple. So we need different code to run forward function.